File: sdks/php/tests/Api/ApprovalApiTest.php
<?php

declare(strict_types=1);

namespace Atap\Sdk\Tests\Api;

use Atap\Sdk\Model\Approval;
use Atap\Sdk\Model\ApprovalSubject;
use Atap\Sdk\Tests\TestCase;

class ApprovalApiTest extends TestCase
{
    public function testCreate(): void
    {
        $client = $this->createAuthedClient([
            $this->jsonResponse(201, [
                'id' => 'apr_test123',
                'state' => 'pending',
                'from' => 'did:web:localhost%3A8080:agent:a',
                'to' => 'did:web:localhost%3A8080:human:b',
                'subject' => [
                    'type' => 'data_access',
                    'label' => 'Access user data',
                    'payload' => ['scope' => 'read'],
                ],
                'created_at' => '2024-01-01T00:00:00Z',
            ]),
        ]);

        $subject = new ApprovalSubject('data_access', 'Access user data', payload: ['scope' => 'read']);
        $approval = $client->approvals()->create(
            'did:web:localhost%3A8080:agent:a',
            'did:web:localhost%3A8080:human:b',
            $subject,
        );

        $this->assertInstanceOf(Approval::class, $approval);
        $this->assertSame('apr_test123', $approval->id);
        $this->assertSame('pending', $approval->state);
        $this->assertNotNull($approval->subject);
        $this->assertSame('data_access', $approval->subject->type);
    }

    public function testRespond(): void
    {
        $client = $this->createAuthedClient([
            $this->jsonResponse(200, [
                'id' => 'apr_test123',
                'state' => 'approved',
                'responded_at' => '2024-01-01T00:01:00Z',
                'signatures' => ['approver' => 'sig_data'],
            ]),
        ]);

        $approval = $client->approvals()->respond('apr_test123', 'jws_signature');
        $this->assertSame('approved', $approval->state);
        $this->assertNotNull($approval->respondedAt);
    }

    public function testList(): void
    {
        $client = $this->createAuthedClient([
            $this->jsonResponse(200, [
                'approvals' => [
                    ['id' => 'apr_1', 'state' => 'pending', 'from' => 'a', 'to' => 'b'],
                    ['id' => 'apr_2', 'state' => 'approved', 'from' => 'c', 'to' => 'd'],
                ],
            ]),
        ]);

        $approvals = $client->approvals()->list();
        $this->assertCount(2, $approvals);
        $this->assertSame('apr_1', $approvals[0]->id);
        $this->assertSame('apr_2', $approvals[1]->id);
    }

    public function testListWithDirectArray(): void
    {
        $client = $this->createAuthedClient([
            $this->jsonResponse(200, [
                'items' => [
                    ['id' => 'apr_1', 'state' => 'pending', 'from' => 'a', 'to' => 'b'],
                ],
            ]),
        ]);

        $approvals = $client->approvals()->list();
        $this->assertCount(1, $approvals);
    }

    public function testRevoke(): void
    {
        $client = $this->createAuthedClient([
            $this->jsonResponse(200, [
                'id' => 'apr_test123',
                'state' => 'revoked',
            ]),
        ]);

        $approval = $client->approvals()->revoke('apr_test123');
        $this->assertSame('revoked', $approval->state);
    }

    public function testCreateWithVia(): void
    {
        $client = $this->createAuthedClient([
            $this->jsonResponse(201, [
                'id' => 'apr_via',
                'state' => 'pending',
                'via' => 'did:web:mediator',
                'from' => 'a',
                'to' => 'b',
            ]),
        ]);

        $subject = new ApprovalSubject('action', 'Test');
        $approval = $client->approvals()->create('a', 'b', $subject, 'did:web:mediator');
        $this->assertSame('did:web:mediator', $approval->via);
    }
}

File: sdks/java/src/test/java/dev/atap/sdk/api/ApprovalApiTest.java
package dev.atap.sdk.api;

import com.github.tomakehurst.wiremock.WireMockServer;
import com.github.tomakehurst.wiremock.core.WireMockConfiguration;
import dev.atap.sdk.ATAPClient;
import dev.atap.sdk.TestFixtures;
import dev.atap.sdk.model.Approval;
import dev.atap.sdk.model.ApprovalSubject;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.Collections;
import java.util.List;

import static com.github.tomakehurst.wiremock.client.WireMock.*;
import static org.junit.jupiter.api.Assertions.*;

class ApprovalApiTest {

    private WireMockServer wireMock;
    private ATAPClient client;

    @BeforeEach
    void setUp() {
        wireMock = new WireMockServer(WireMockConfiguration.wireMockConfig().dynamicPort());
        wireMock.start();

        wireMock.stubFor(post(urlEqualTo("/v1/oauth/token"))
                .willReturn(aResponse()
                        .withStatus(200)
                        .withHeader("Content-Type", "application/json")
                        .withBody(TestFixtures.tokenResponseJson())));

        client = ATAPClient.builder()
                .baseUrl("http://localhost:" + wireMock.port())
                .did(TestFixtures.TEST_DID)
                .keyPair(TestFixtures.getTestKeyPair())
                .clientSecret(TestFixtures.TEST_CLIENT_SECRET)
                .platformDomain(TestFixtures.TEST_PLATFORM_DOMAIN)
                .build();
    }

    @AfterEach
    void tearDown() {
        wireMock.stop();
        client.close();
    }

    @Test
    void testCreate() {
        wireMock.stubFor(post(urlEqualTo("/v1/approvals"))
                .willReturn(aResponse()
                        .withStatus(200)
                        .withHeader("Content-Type", "application/json")
                        .withBody(TestFixtures.approvalResponseJson())));

        ApprovalSubject subject = new ApprovalSubject("data_access", "Access user data", true, null);
        Approval approval = client.approvals.create("did:web:example:agent:a",
                "did:web:example:agent:b", subject);

        assertNotNull(approval);
        assertEquals("apr_abc123", approval.getId());
        assertEquals("pending", approval.getState());
        assertEquals("did:web:example:agent:a", approval.getFromDid());
        assertEquals("did:web:example:agent:b", approval.getToDid());
        assertNotNull(approval.getSubject());
        assertEquals("data_access", approval.getSubject().getType());
    }

    @Test
    void testCreateWithVia() {
        wireMock.stubFor(post(urlEqualTo("/v1/approvals"))
                .willReturn(aResponse()
                        .withStatus(200)
                        .withHeader("Content-Type", "application/json")
                        .withBody(TestFixtures.approvalResponseJson())));

        ApprovalSubject subject = new ApprovalSubject("data_access", "Access user data", false,
                Collections.singletonMap("resource", "users"));
        Approval approval = client.approvals.create("did:web:example:agent:a",
                "did:web:example:agent:b", subject, "did:web:example:machine:proxy");

        assertNotNull(approval);
    }

    @Test
    void testRespond() {
        String respondedJson = "{\"id\":\"apr_abc123\","
                + "\"state\":\"approved\","
                + "\"from\":\"did:web:example:agent:a\","
                + "\"to\":\"did:web:example:agent:b\","
                + "\"subject\":{\"type\":\"data_access\",\"label\":\"Access\"},"
                + "\"responded_at\":\"2025-01-01T01:00:00Z\","
                + "\"signatures\":{\"approver\":\"sig_data\"}}";

        wireMock.stubFor(post(urlEqualTo("/v1/approvals/apr_abc123/respond"))
                .willReturn(aResponse()
                        .withStatus(200)
                        .withHeader("Content-Type", "application/json")
                        .withBody(respondedJson)));

        Approval approval = client.approvals.respond("apr_abc123", "sig_data");
        assertNotNull(approval);
        assertEquals("approved", approval.getState());
        assertNotNull(approval.getRespondedAt());
    }

    @Test
    void testList() {
        String listJson = "{\"approvals\":[" + TestFixtures.approvalResponseJson() + "]}";

        wireMock.stubFor(get(urlEqualTo("/v1/approvals"))
                .willReturn(aResponse()
                        .withStatus(200)
                        .withHeader("Content-Type", "application/json")
                        .withBody(listJson)));

        List<Approval> approvals = client.approvals.list();
        assertNotNull(approvals);
        assertEquals(1, approvals.size());
        assertEquals("apr_abc123", approvals.get(0).getId());
    }

    @Test
    void testRevoke() {
        wireMock.stubFor(delete(urlEqualTo("/v1/approvals/apr_abc123"))
                .willReturn(aResponse()
                        .withStatus(200)
                        .withHeader("Content-Type", "application/json")
                        .withBody("{\"id\":\"apr_abc123\",\"state\":\"revoked\"}")));

        Approval approval = client.approvals.revoke("apr_abc123");
        assertNotNull(approval);
        assertEquals("revoked", approval.getState());
    }
}
